src/music_gui.cpp

```diff
@@ -35,278 +36,388 @@
-	for (uint k = 0; k < NUM_SONG_CLASSES; k++) {
-		j = 0;
-		for (uint i = 0; i < NUM_SONGS_CLASS; i++) {
-			int id = k * NUM_SONGS_CLASS + i + 1;
-			if (StrEmpty(GetSongName(id))) continue;
-			_playlists[k + 1][j++] = id + 1;
-		}
-		/* Terminate the list */
-		_playlists[k + 1][j] = 0;
+	_settings_client.music.shuffle = true;
+
+	this->active_playlist = this->displayed_playlist;
+	for (size_t i = 0; i < this->active_playlist.size(); i++) {
+		size_t shuffle_index = InteractiveRandom() % (this->active_playlist.size() - i);
+		auto &a = this->active_playlist[i];
+		auto &b = this->active_playlist[i + shuffle_index];
+		std::swap(a, b);
 	}
 
-	ValidatePlaylist(_settings_client.music.custom_1, lastof(_settings_client.music.custom_1));
-	ValidatePlaylist(_settings_client.music.custom_2, lastof(_settings_client.music.custom_2));
+	if (_settings_client.music.playing) this->Play();
 
-	if (BaseMusic::GetUsedSet()->num_available < _music_wnd_cursong) {
-		/* If there are less songs than the currently played song,
-		 * just pause and reset to no song. */
-		_music_wnd_cursong = 0;
-		_song_is_active = false;
-	}
+	InvalidateWindowData(WC_MUSIC_WINDOW, 0);
 }
 
-static void SkipToPrevSong()
+/** Disable shuffle and restart playback */
+void MusicSystem::Unshuffle()
 {
-	byte *b = _cur_playlist;
-	byte *p = b;
-	byte t;
+	_settings_client.music.shuffle = false;
+	this->active_playlist = this->displayed_playlist;
 
-	if (b[0] == 0) return; // empty playlist
+	if (_settings_client.music.playing) this->Play();
 
-	do p++; while (p[0] != 0); // find the end
+	InvalidateWindowData(WC_MUSIC_WINDOW, 0);
+}
 
-	t = *--p; // and copy the bytes
-	while (p != b) {
-		p--;
-		p[1] = p[0];
-	}
-	*b = t;
+/** Start/restart playback at current song */
+void MusicSystem::Play()
+{
+	/* Always set the playing flag, even if there is no music */
+	_settings_client.music.playing = true;
+	MusicDriver::GetInstance()->StopSong();
+	this->ChangePlaylistPosition(0);
+
+	/* If there is no music, don't try to play it */
+	if (this->active_playlist.empty()) return;
 
-	_song_is_active = false;
+	MusicSongInfo song = this->active_playlist[this->playlist_position];
+	if (_game_mode == GM_MENU && this->selected_playlist == PLCH_THEMEONLY) song.loop = true;
+	MusicDriver::GetInstance()->PlaySong(song);
+
+	InvalidateWindowData(WC_MUSIC_WINDOW, 0);
 }
 
-static void SkipToNextSong()
+/** Stop playback and set flag that we don't intend to play music */
+void MusicSystem::Stop()
 {
-	byte *b = _cur_playlist;
-	byte t;
-
-	t = b[0];
-	if (t != 0) {
-		while (b[1] != 0) {
-			b[0] = b[1];
-			b++;
-		}
-		b[0] = t;
-	}
+	MusicDriver::GetInstance()->StopSong();
+	_settings_client.music.playing = false;
 
-	_song_is_active = false;
+	InvalidateWindowData(WC_MUSIC_WINDOW, 0);
 }
 
-static void MusicVolumeChanged(byte new_vol)
+/** Skip to next track */
+void MusicSystem::Next()
 {
-	MusicDriver::GetInstance()->SetVolume(new_vol);
+	this->ChangePlaylistPosition(+1);
```

Comment:
I think this is kinda neat for readability actually, does absolutely nothing, but clearly shows what's going on :)